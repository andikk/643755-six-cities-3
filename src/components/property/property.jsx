import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import ReviewsList from "../reviews-list/reviews-list.jsx";
import Map from "../map/map.jsx";
import PlacesList from "../places-list/places-list.jsx";
import {connect} from "react-redux";
import {
  getOfferByIdSelector,
  getReviewsSelector,
  getOffersNearbySelector,
  getAuthorizationStatusSelector,
  getReviewsCountSelector
} from "../../selectors";
import ReviewForm from "../review-form/review-form.jsx";
import {Operation} from "../../reducer.js";
import UserNav from "../user-nav/user-nav.jsx";
import {BookmarkActions} from "../../const.js";

// страница предложения
class Property extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidUpdate(prevProps) {
    if (this.props.card !== prevProps.card) {
      window.scrollTo(0, 0);
    }
  }

  componentDidMount() {
    const {loadReviews, loadNearby, offerId} = this.props;
    loadReviews(offerId);
    loadNearby(offerId);
  }

  render() {
    if (!this.props.card) {
      return null;
    }

    const {addToFavorite, offerId, addComment, location, reviewsCount} = this.props;

    const handleBookmarkClick = () => {

      const {ADD, REMOVE} = BookmarkActions;
      const status = isFavorite ? REMOVE : ADD;

      addToFavorite(offerId, status);
    };

    const {card = {}, reviews = [], neighborhood = [], authorizationStatus} = this.props;
    const {photos, description, premium, bedrooms, guests, features, owner, price, rating, ratingValue, name, type, city, coordinates, isFavorite} = card;
    const coordinatesNearby = neighborhood.map((item) => (item.coordinates));

    return (
      <div className="page">
        <header className="header">
          <div className="container">
            <div className="header__wrapper">
              <div className="header__left">
                <a className="header__logo-link" href="/">
                  <img className="header__logo" src="img/logo.svg" alt="6 cities logo" width="81" height="41"/>
                </a>
              </div>
              <UserNav referer={location.pathname}/>
            </div>
          </div>
        </header>

        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">

                {photos.filter((photo, index) => index < 6)
                  .map((photo, index) => (
                    <div className="property__image-wrapper" key={photo + index}>
                      <img className="property__image" src={photo} alt={name}/>
                    </div>
                  ))}
              </div>

            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                {premium &&
                <div className="property__mark">
                  <span>Premium</span>
                </div>
                }

                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {name}
                  </h1>
                  <button
                    style={{position: `absolute`, top: `41px`, right: `93px`}}
                    className={`place-card__bookmark-button place-card__bookmark-button${isFavorite ? `--active` : ``} button`}
                    type="button"
                    onClick={handleBookmarkClick}>
                    <svg className="place-card__bookmark-icon" width="31" height="33">
                      <use xlinkHref="#icon-bookmark"></use>
                    </svg>
                    <span className="visually-hidden">In bookmarks</span>
                  </button>
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: `${ratingValue <= 4 ? ratingValue * 20 : 100}%`}}/>
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value">{rating}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">
                    {type}
                  </li>
                  <li className="property__feature property__feature--bedrooms">
                    {bedrooms} Bedrooms
                  </li>
                  <li className="property__feature property__feature--adults">
                    Max {guests} adults
                  </li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">&euro;{price}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <ul className="property__inside-list">
                    {features.map((feature, index) => (
                      <li className="property__inside-item" key={feature + index}>
                        {feature}
                      </li>
                    ))}
                  </ul>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div
                      className={`property__avatar-wrapper ${owner.super ? ` property__avatar-wrapper--pro` : ``} user__avatar-wrapper`}>
                      <img className="property__avatar user__avatar" src={owner.src} width="74" height="74"
                        alt={owner.name}/>
                    </div>
                    <span className="property__user-name">
                      {owner.name}
                    </span>
                  </div>
                  <div className="property__description">
                    <p className="property__text">
                      {description}
                    </p>
                  </div>
                </div>
                <section className="property__reviews reviews">
                  <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviewsCount}</span></h2>
                  <ReviewsList className="property__reviews" reviews={reviews}/>
                  {(authorizationStatus === `AUTH`) &&
                  <ReviewForm offerId={offerId} onSubmit={addComment}/>
                  }
                </section>

              </div>
            </div>

            <Map className="property__map"
              city={city}
              coordinates={[...coordinatesNearby, coordinates]}
              activeMarker={coordinates}/>

          </section>
          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>

              <PlacesList offers={neighborhood}
                className="near-places__list places__list"
                classNameForArticle="near-places__card"
                imgSize={{width: 260, height: 200}}
                classNameForItems="near-places"/>

            </section>
          </div>
        </main>
      </div>
    );
  }
}

Property.defaultProps = {
  location: {}
};

Property.propTypes = {
  card: PropTypes.shape({
    photos: PropTypes.array,
    description: PropTypes.string,
    premium: PropTypes.bool,
    bedrooms: PropTypes.number,
    guests: PropTypes.number,
    features: PropTypes.array,
    owner: PropTypes.object,
    price: PropTypes.number,
    rating: PropTypes.number,
    ratingValue: PropTypes.number,
    name: PropTypes.string,
    type: PropTypes.string,
    city: PropTypes.object,
    coordinates: PropTypes.array
  }),
  loadReviews: PropTypes.func,
  loadNearby: PropTypes.func,
  offerId: PropTypes.number,
  reviews: PropTypes.array,
  neighborhood: PropTypes.array,
  onCardHover: PropTypes.func,
  addToFavorite: PropTypes.func,
  authorizationStatus: PropTypes.string,
  addComment: PropTypes.func,
  location: PropTypes.object,
  reviewsCount: PropTypes.number
};

const mapStateToProps = (state, ownProps) => {
  const id = Number(ownProps.match.params.id);
  return ({
    card: getOfferByIdSelector(state)(id),
    offerId: Number(ownProps.match.params.id),
    reviews: getReviewsSelector(state),
    neighborhood: getOffersNearbySelector(state),
    authorizationStatus: getAuthorizationStatusSelector(state),
    reviewsCount: getReviewsCountSelector(state)
  });
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  loadReviews: () => dispatch(Operation.loadReviews(ownProps.match.params.id)),
  loadNearby: () => dispatch(Operation.loadNearby(ownProps.match.params.id)),
  addToFavorite: (id, status) => {
    return dispatch(Operation.addToFavorite(id, status))
      .catch(() => {
        ownProps.history.push(`/login`, {referer: `/offer/${id}`});
      });
  },
  addComment: (commentData, offerId) => dispatch(Operation.addComment(commentData, offerId))
});

export {Property};
export default connect(mapStateToProps, mapDispatchToProps)(Property);
