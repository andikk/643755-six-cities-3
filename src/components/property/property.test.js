import React from "react";
import renderer from "react-test-renderer";
import {Property} from "./property.jsx";
import createAPI from '../../api';
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";
import thunk from "redux-thunk";

const api = createAPI();
const mockStore = configureStore([thunk.withExtraArgument(api)]);

it(`Should Property render correctly`, () => {
  const store = mockStore({
    city: {name: `Cologne`, location: {"latitude": 50.938361, "longitude": 6.959974, "zoom": 13}},
    offersIds: [1],
    offersMap: {
      1: {id: 1,
        city: {name: `Cologne`, location: {"latitude": 50.938361, "longitude": 6.959974, "zoom": 13}},
        src: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg`,
        photos: [`https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/10.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/11.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/12.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/13.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/8.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/1.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/6.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/20.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/7.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/5.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/18.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/14.jpg`],
        description: `Design interior in most sympathetic area! Complitely renovated, well-equipped, cosy studio in idyllic, over 100 years old wooden house. Calm street, fast connection to center and airport.`,
        premium: false,
        bedrooms: 5,
        guests: 5,
        features: [`Breakfast`, `Washer`, `Air conditioning`, `Laptop friendly workspace`, `Baby seat`],
        owner: {id: 25, name: `Angelina`, super: true, src: `img/avatar-angelina.jpg`},
        price: 403,
        rating: 2.9,
        name: `Wood and stone place`,
        type: `house`,
        coordinates: [50.932361, 6.937974],
        reviews: [],
        neighborhood: [],
        isFavorite: false}
    },
    offersNearbyIds: [],
    offersFavoritesIds: [],
    reviews: [],
    activeOfferId: null,
    authorizationStatus: `AUTH`,
    user: {
      avatar: `img/1.png`,
      id: 4,
      isTrust: false,
      name: `Max`
    },
    activeFilter: {label: `Popular`, value: `ALL`}
  });

  const card = {id: 1,
    city: {name: `Cologne`, location: {"latitude": 50.938361, "longitude": 6.959974, "zoom": 13}},
    src: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg`,
    photos: [`https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/10.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/11.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/12.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/13.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/8.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/1.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/6.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/20.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/7.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/5.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/18.jpg`, `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/14.jpg`],
    description: `Design interior in most sympathetic area! Complitely renovated, well-equipped, cosy studio in idyllic, over 100 years old wooden house. Calm street, fast connection to center and airport.`,
    premium: false,
    bedrooms: 5,
    guests: 5,
    features: [`Breakfast`, `Washer`, `Air conditioning`, `Laptop friendly workspace`, `Baby seat`],
    owner: {id: 25, name: `Angelina`, super: true, src: `img/avatar-angelina.jpg`},
    price: 403,
    rating: 2.9,
    name: `Wood and stone place`,
    type: `house`,
    coordinates: [50.932361, 6.937974],
    reviews: [],
    neighborhood: [],
    isFavorite: false};

  const tree = renderer
    .create(
        <Provider store={store}>
          <MemoryRouter>
            <Property
              match={{params: 1}}
              card={card}
              loadReviews={()=>{}}
              loadNearby={()=>{}}
              offerId={1}
              reviews={[]}
              neighborhood={[]}
              onCardHover={()=>{}}
              addToFavorite={()=>{}}
              authorizationStatus={`PropTypes.string`}
              addComment={()=>{}}/>
          </MemoryRouter>
        </Provider>
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
