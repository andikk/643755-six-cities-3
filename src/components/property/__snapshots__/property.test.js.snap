// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Property render correctly 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                >
                   
                </div>
                <span
                  className="header__user-name user__name"
                />
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Wood and stone place"
              className="property__image"
              src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/10.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Wood and stone place"
              className="property__image"
              src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/11.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Wood and stone place"
              className="property__image"
              src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/12.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Wood and stone place"
              className="property__image"
              src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/13.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Wood and stone place"
              className="property__image"
              src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/8.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Wood and stone place"
              className="property__image"
              src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Wood and stone place
            </h1>
            <button
              className="place-card__bookmark-button place-card__bookmark-button button"
              onClick={[Function]}
              style={
                Object {
                  "position": "absolute",
                  "right": "93px",
                  "top": "41px",
                }
              }
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height="33"
                width="31"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                In bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              2.9
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              house
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              5
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              5
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              403
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Breakfast
              </li>
              <li
                className="property__inside-item"
              >
                Washer
              </li>
              <li
                className="property__inside-item"
              >
                Air conditioning
              </li>
              <li
                className="property__inside-item"
              >
                Laptop friendly workspace
              </li>
              <li
                className="property__inside-item"
              >
                Baby seat
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper  property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Angelina"
                  className="property__avatar user__avatar"
                  height="74"
                  src="img/avatar-angelina.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                Design interior in most sympathetic area! Complitely renovated, well-equipped, cosy studio in idyllic, over 100 years old wooden house. Calm street, fast connection to center and airport.
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews · 
              <span
                className="reviews__amount"
              />
            </h2>
            <ul
              className="reviews__list"
            />
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      />
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          className="near-places__list places__list"
        />
      </section>
    </div>
  </main>
</div>
`;
